program/tokenprog/state.go

```diff
@@ -13,13 +13,52 @@ const MultisigAccountSize uint64 = 355
 const MintAccountSize = 82
 
 type MintAccount struct {
-	MintAuthorityOption   uint32
-	MintAuthority         common.PublicKey
-	Supply                uint64
-	Decimals              uint8
-	IsInitialized         bool
-	FreezeAuthorityOption uint32
-	FreezeAuthority       common.PublicKey
+	// MintAuthorityOption   uint32
+	MintAuthority *common.PublicKey
+	Supply        uint64
+	Decimals      uint8
+	IsInitialized bool
+	// FreezeAuthorityOption uint32
+	FreezeAuthority *common.PublicKey
+}
+
+const (
+	mintMintAuthorityOptionOffset   = 0
+	mintMintAuthorityOffset         = mintMintAuthorityOptionOffset + 4
+	mintSupplyOffset                = mintMintAuthorityOffset + 32
+	mintDecimalsOffset              = mintSupplyOffset + 8
+	mintIsInitializedOffset         = mintDecimalsOffset + 1
+	mintFreezeAuthorityOptionOffset = mintIsInitializedOffset + 1
+	mintFreezeAuthorityOffset       = mintFreezeAuthorityOptionOffset + 4
+)
+
+func isSome(option []byte) bool {
+	return bytes.Equal(option, Some)
+}
+
+func MintAccountFromData(data []byte) (*MintAccount, error) {
+	if len(data) != MintAccountSize {
+		return nil, fmt.Errorf("mint account data length mismatch")
+	}
+
+	var mint MintAccount
+
+	mintAuthorityOption := data[0:4]
+	if isSome(mintAuthorityOption) {
+		key := common.PublicKeyFromBytes(data[mintMintAuthorityOffset : mintSupplyOffset+32])
+		mint.MintAuthority = &key
+	}
+
+	mint.Supply = binary.LittleEndian.Uint64(data[mintSupplyOffset : mintSupplyOffset+8])
+	mint.Decimals = uint8(data[mintDecimalsOffset])
+	mint.IsInitialized = data[mintIsInitializedOffset] == 1
+
+	if isSome(data[mintFreezeAuthorityOptionOffset:mintFreezeAuthorityOptionOffset+4]) {
+		key := common.PublicKeyFromBytes(data[mintFreezeAuthorityOffset : mintFreezeAuthorityOffset+32])
+		mint.FreezeAuthority = &key
+	}
+
+	return &mint, nil
```

Comment:
I guess maybe you want to make code more readable but I would like to use literal number like I decode token account and would you add a test for this?

Reply:
how important is this to you to use literal number? using literal offsets is pretty unreadable to be honest.

Reply:
I want to use literal number for some reasons.

1. if someday I want to use get accountInfo to fetch data and using offset. the literal number is more intuitive to me. I don't need to recalculate it if I miss in the bytes array.
2. I don't think there are someone will read this code, I want to make them as clear as possible.

Reply:
> if someday I want to use get accountInfo to fetch data and using offset.

if that's the case, wouldn't you want to export the offset constants rather than using literals?